Packages/ClientRuntime/Sources/Serialization/SerializationUtils/TimestampSerdeUtils.swift

```diff
@@ -0,0 +1,274 @@
+    /// For each format, the string will only contain fractional seconds if a non-zero value exists for fractional seconds.
+    ///
+    /// - Parameter date: The date to be represented.
+    /// - Returns: A smithy timestamp formatted string representing the date.
+    public func string(from date: Date) -> String {
+        switch format {
+        case .epochSeconds:
+            let seconds = date.timeIntervalSince1970
+            return date.hasFractionalSeconds
+            ? String(seconds)
+            : String(format: "%.0f", seconds)
+        case .dateTime:
+            return date.hasFractionalSeconds
+            ? date.iso8601WithFractionalSeconds()
+            : date.iso8601WithoutFractionalSeconds()
+        case .httpDate:
+            return date.hasFractionalSeconds
+            ? date.rfc5322WithFractionalSeconds()
+            : date.rfc5322WithoutFractionalSeconds()
+        }
+    }
+    
+    /// Creates and returns a date object from the specified smithy timestamp string representation.
+    ///
+    /// - Parameter string: The smithy timestamp formatted string representation of a date.
+    /// - Returns: A date object, or nil if no valid date was found.
+    public func date(from string: String) -> Date? {
+        // Fractional seconds may be optionally included
+        // therfore we need to attempt to get the date using both formatters (with fractional seconds and without)
+        switch format {
+        case .epochSeconds:
+            return Double(string).map(Date.init(timeIntervalSince1970:))
+        case .dateTime:
+            return Date(
+                from: string,
+                formatters: [
+                    .iso8601DateFormatterWithFractionalSeconds,
+                    .iso8601DateFormatterWithoutFractionalSeconds
+                ]
+            )
+        case .httpDate:
+            return Date(
+                from: string,
+                formatters: [
+                    .rfc5322WithFractionalSeconds,
+                    .rfc5322WithoutFractionalSeconds
+                ]
+            )
+        }
+    }
+}
+
+// MARK: - Encoding Helpers
+
+/// A struct to encapsulate the encoding logic for Timestamps
+struct TimestampEncodable: Encodable {
+    let date: Date
+    let format: TimestampFormat
+    
+    init(date: Date, format: TimestampFormat) {
+        self.date = date
+        self.format = format
+    }
+    
+    /// Encodes the date according to the format.
+    func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+        switch format {
+        // We want epoch seconds to be encoded as a Double and not a string
+        case .epochSeconds:
+            try container.encode(date.timeIntervalSince1970)
+        default:
+            let dateString = TimestampFormatter(format: format).string(from: date)
+            try container.encode(dateString)
+        }
+    }
+}
+
+extension KeyedEncodingContainer {
```

Comment:
how does  the encoding decoding works for the list types used in header and query params?

Reply:
I will investigate this today to ensure the only change here is that it might contain fractional seconds (if the date includes non-zero fractional seconds).

Reply:
These cases are covered by protocol tests and are passing